kernels/volk/volk_32fc_x2_s32fc_multiply_conjugate_32fc.h

```diff
@@ -0,0 +1,247 @@
+ *
+ * GNU Radio is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * GNU Radio is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Radio; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+/*!
+ * \page volk_32fc_x2_s32fc_multiply_conjugate_32fc
+ *
+ * \b Overview
+ *
+ * Conjugate the input complex vector, multiply them by a complex scalar,
+ * add the another input complex vector and returns the results.
+ *
+ * c[i] = a[i] + conj(b[i]) * scalar
+ *
+ * <b>Dispatcher Prototype</b>
+ * \code
+ * void volk_32fc_x2_s32fc_multiply_conjugate_32fc(lv_32fc_t* cVector, const lv_32fc_t* aVector, const lv_32fc_t* bVector, const lv_32fc_t scalar, unsigned int num_points);
+ * \endcode
+ *
+ * \b Inputs
+ * \li aVector: The input vector to be added.
+ * \li bVector: The input vector to be conjugate and multiplied.
+ * \li scalar The complex scalar to multiply against conjugated bVector.
+ * \li num_points: The number of complex values in aVector.
+ *
+ * \b Outputs
+ * \li cVector: The vector where the results will be updated.
+ *
+ * \b Example
+ *
+ * \code
+ *
+ * \endcode
+ */
+
+#ifndef INCLUDED_volk_32fc_x2_s32fc_multiply_conjugate_32fc_a_H
+#define INCLUDED_volk_32fc_x2_s32fc_multiply_conjugate_32fc_a_H
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <volk/volk_complex.h>
+#include <float.h>
+
+
+#ifdef LV_HAVE_AVX
+#include <immintrin.h>
+
+static inline void volk_32fc_x2_s32fc_multiply_conjugate_32fc_a_avx(lv_32fc_t* cVector, const lv_32fc_t* aVector, const lv_32fc_t* bVector, const lv_32fc_t scalar, unsigned int num_points) {
+    unsigned int number = 0;
+    unsigned int i = 0;
+    const unsigned int quarterPoints = num_points / 4;
+    unsigned int isodd = num_points & 3;
+
+    __m256 x, sl, sh, y, z, tmp1, tmp2;
+
+    const lv_32fc_t* a = aVector;
+    const lv_32fc_t* b = bVector;
+    lv_32fc_t* c = cVector;
+
+    __m256 conjugator = _mm256_setr_ps(0, -0.f, 0, -0.f, 0, -0.f, 0, -0.f);
+
+    // Set up constant scalar vector
+    sl = _mm256_set1_ps(lv_creal(scalar));
+    sh = _mm256_set1_ps(lv_cimag(scalar));
+
+    for(;number < quarterPoints; number++) {
+        x = _mm256_load_ps((float*)b);
+        y = _mm256_load_ps((float*)a);
+
+        x = _mm256_xor_ps(x, conjugator);
+
+        tmp1 = _mm256_mul_ps(x,sl);
+
+        x = _mm256_shuffle_ps(x,x,0xB1);
+
+        tmp2 = _mm256_mul_ps(x,sh);
+
+        z = _mm256_addsub_ps(tmp1,tmp2);
+
+        z = _mm256_add_ps(y, z);
```

Comment:
Could you move these lines to `include/volk_avx_intrinsics.h`? Also, that would make it easier to add an unaligned version too. If you could add an unaligned version, that would be great.

Reply:
Thank you for reviewing. I have a question.

"move these lines to include/volk_avx_intrinsics.h?" means:
1. "add method to include/volk_avx_intrinsics.h", for EXAMPLE:
static inline __m256 _mm256_complexconjugatemulscalar_ps(__m256 x, __m256 y, __m256 sl, __m256 sh) {
   __m256 z, tmp1, tmp2;
   __m256 conjugator = _mm256_setr_ps(0, -0.f, 0, -0.f, 0, -0.f, 0, -0.f);
   x = _mm256_xor_ps(x, conjugator);
   tmp1 = _mm256_mul_ps(x,sl);
   x = _mm256_shuffle_ps(x,x,0xB1);
   tmp2 = _mm256_mul_ps(x,sh);
   z = _mm256_addsub_ps(tmp1,tmp2);
   return _mm256_add_ps(y, z);
}

2. and "volk_fc32_x2_s32fc_multiply_conjugate_32fc_a_avx()?", for EXAMPLE:
...
for (;number < quarterPoints; number++) {
    x = _mm256_load_ps((float*)b);
    y = _mm256_load_ps((float*)a);
    z = _mm256_complexconjugatemulscalar_ps(x, y, sl, sh);
    _mm256_store_ps((float*)c,z);
...

Reply:
Yes. that's it. There is already some code for conjugation and multiplication. I assume you can at least use the conjugate function. Maybe you can reuse complex multiplication as well. Though, I suggest to not force your code to use this function. only if it is a one-to-one replacement like in the conjugator case.

Reply:
I might find this can be realized by combinations of existing intrinsics... I'm studying now.
Thanks for your suggestion.